File: apps/api/app/orchestrator/intake_graph.py
"""Intake-phase LangGraph (V32.M2.3.4 / F-320 + F-321).

Independent graph from `turn_graph.py`. The two never interleave — turn_graph
runs per-turn during the live interview, this one runs once during intake
to fan out Parse + Research in parallel and (optionally) hand the merged
inputs to Framework Agent for predicted-question generation.

  START ──► parse_node    ───┐
       │                     ├──► predict_questions_node ──► END
       └──► research_node ───┘

Parallelism rules:
  * `parse_node` and `research_node` are siblings off START. LangGraph runs
    them concurrently and the merge into `predict_questions_node` waits for
    both to commit their state slice.
  * `research_node` is "soft" — when the user has NOT opted in (or the
    Research call errors / times out) it sets `research_payload=None` and
    flips `research_skipped=True`. This NEVER blocks parse_node; it's a
    no-op write into state.
  * `predict_questions_node` requires a parse_payload to do anything
    useful. When parse_payload is missing or framework_config is None we
    skip the FrameworkAgent call and leave `direction_framework=None`.

L0 A11 privacy is inherited from the agents' own schemas/prompts —
intake_graph itself never sees resume_text after the parse_node populates
parse_payload. The research_input is built by the caller (assets domain
service) using ONLY company / role / industry hints from the JD.
"""
from __future__ import annotations

import asyncio
import logging
from typing import Any

from langgraph.graph import END, START, StateGraph
from pydantic import BaseModel

from app.agents.framework.schemas import FrameworkAgentInput, FrameworkAgentOutput, FrameworkConfigInput
from app.agents.framework.service import FrameworkAgentService
from app.agents.parse.schemas import ParseAgentInput
from app.agents.parse.service import ParseAgentService
from app.agents.research.schemas import ResearchAgentInput, ResearchAgentOutput
from app.agents.research.service import ResearchAgentService
from app.infra.llm import LLMGateway
from app.infra.llm.errors import LLMError
from app.schemas.parse import ParseResultPayload

logger = logging.getLogger(__name__)


# Spec A7-style lock: any rename / add / remove drift here must be
# matched by the test_intake_graph_contract.py assertion.
INTAKE_GRAPH_NODES: frozenset[str] = frozenset(
    {"parse_node", "research_node", "predict_questions_node"}
)

# Soft cap so a hung Research call cannot block parse + framework forever.
RESEARCH_TIMEOUT_SECONDS: float = 15.0


class IntakeState(BaseModel):
    """State threaded through intake_graph. Mirrors TurnState style."""

    # ===== Inputs (set before graph.ainvoke) =====
    resume_text: str
    jd_text: str
    # research_input None ⇒ user has NOT opted in (or the caller could
    # not derive company/role/industry hints). research_node short-
    # circuits to a no-op.
    research_input: ResearchAgentInput | None = None
    # framework_config None ⇒ predict_questions_node is a no-op. The
    # session-creation flow supplies a config; the parse-trigger flow
    # leaves it None and just gets parse + research back.
    framework_config: FrameworkConfigInput | None = None

    # ===== Outputs (filled by nodes) =====
    parse_payload: ParseResultPayload | None = None
    research_payload: ResearchAgentOutput | None = None
    research_skipped: bool = False
    research_error: str | None = None
    direction_framework: FrameworkAgentOutput | None = None


def build_intake_graph(
    gateway: LLMGateway,
    *,
    parse_service: ParseAgentService | None = None,
    research_service: ResearchAgentService | None = None,
    framework_service: FrameworkAgentService | None = None,
    research_timeout: float = RESEARCH_TIMEOUT_SECONDS,
):
    """Compile the 3-node intake StateGraph.

    Service overrides are exposed for tests so they can inject scripted
    fakes (see test_intake_graph_contract.py). Production callers pass
    just the gateway and rely on default singletons.
    """
    parse_service = parse_service or ParseAgentService()
    research_service = research_service or ResearchAgentService()
    framework_service = framework_service or FrameworkAgentService()

    async def parse_node(state: IntakeState) -> dict[str, Any]:
        agent_output = await parse_service.run(
            ParseAgentInput(resume_text=state.resume_text, jd_text=state.jd_text),
            gateway,
        )
        # The parse agent's output IS-A ParseResultPayload (subclass);
        # we project it onto the parent type so downstream consumers
        # can serialize without dragging the agent module in.
        payload = ParseResultPayload.model_validate(agent_output.model_dump())
        return {"parse_payload": payload}

    async def research_node(state: IntakeState) -> dict[str, Any]:
        # Opt-out / no input ⇒ soft skip. NEVER block parse_node.
        if state.research_input is None:
            logger.info("research_node_skipped", extra={"reason": "no_input"})
            return {
                "research_payload": None,
                "research_skipped": True,
                "research_error": None,
            }
        try:
            research_output = await asyncio.wait_for(
                research_service.run(state.research_input, gateway),
                timeout=research_timeout,
            )
        except asyncio.TimeoutError:
            logger.info("research_node_timeout")
            return {
                "research_payload": None,
                "research_skipped": True,
                "research_error": "timeout",
            }
        except LLMError as exc:
            logger.info(
                "research_node_llm_error",
                extra={"reason": type(exc).__name__},
            )
            return {
                "research_payload": None,
                "research_skipped": True,
                "research_error": type(exc).__name__,
            }
        return {
            "research_payload": research_output,
            "research_skipped": False,
            "research_error": None,
        }

    async def predict_questions_node(state: IntakeState) -> dict[str, Any]:
        # Skip when we lack either the framework config (parse-trigger
        # path, not session creation) or a parse payload (parse failed
        # before this node ran). Either case leaves direction_framework
        # at None so the upstream session-creation step can run Framework
        # later with full config.
        if state.framework_config is None or state.parse_payload is None:
            return {"direction_framework": None}

        research_json: str | None = None
        if state.research_payload is not None:
            research_json = state.research_payload.model_dump_json()

        framework_input = FrameworkAgentInput(
            parse_payload_json=state.parse_payload.model_dump_json(),
            config=state.framework_config,
            research_payload_json=research_json,
        )
        try:
            framework_output = await framework_service.run(framework_input, gateway)
        except LLMError as exc:
            # Framework failure is upstream-recoverable: session creation
            # already has its own retry+fallback (see SessionsService).
            # We surface None so the caller can re-run Framework directly.
            logger.info(
                "predict_questions_node_framework_error",
                extra={"reason": type(exc).__name__},
            )
            return {"direction_framework": None}
        return {"direction_framework": framework_output}

    graph = StateGraph(IntakeState)
    graph.add_node("parse_node", parse_node)
    graph.add_node("research_node", research_node)
    graph.add_node("predict_questions_node", predict_questions_node)

    # Parallel fan-out from START.
    graph.add_edge(START, "parse_node")
    graph.add_edge(START, "research_node")
    # Both must commit before predict_questions_node fires.
    graph.add_edge("parse_node", "predict_questions_node")
    graph.add_edge("research_node", "predict_questions_node")
    graph.add_edge("predict_questions_node", END)

    return graph.compile()


__all__ = [
    "INTAKE_GRAPH_NODES",
    "RESEARCH_TIMEOUT_SECONDS",
    "IntakeState",
    "build_intake_graph",
]

File: apps/desktop/src/core/graphs/intakeGraph.ts
/**
 * Intake-phase LangGraph (V32.M2.3.4 / F-320 + F-321 + M2.3.X audit fix).
 *
 * Independent graph from `turnGraph.ts`. The two never interleave — turnGraph
 * runs per-turn during the live interview, this one runs once during intake
 * to chain Parse → Research → Framework predict so that downstream UI can
 * show the company / industry / predicted-question cards.
 *
 *   START ──► parse_node ──► research_node ──► predict_questions_node ──► END
 *
 * Sequential rationale (M2.3.X audit-fix):
 *   Research / Predict need company + role + industry-hint signals out of
 *   the JD. Pre-audit we tried to extract these via a regex hand-rolled
 *   inside the assets domain service, which silently always returned None
 *   — so research_node never had a real input and the Research Agent was
 *   never actually called in production. Parse Agent already reads the JD
 *   end-to-end; we now ask it to emit `jd_company_name / jd_role_title /
 *   jd_industry_hints` alongside the rest of its output. research_node
 *   consumes `state.parse_payload.jd_*` to construct its own input, which
 *   forces this to be sequential — research_node MUST run after parse_node
 *   has committed parse_payload.
 *
 * ─── L0 #13 Node-name lock ────────────────────────────────────────────────────
 * The literal triple {"parse_node", "research_node", "predict_questions_node"}
 * is the ONLY allowed spelling. Three-layer enforcement:
 *   ① IntakeGraphNodeName Literal type union (below)
 *   ② INTAKE_GRAPH_NODES = Object.freeze(new Set([...])) runtime constant
 *   ③ Contract test asserts set equality + size 3 + frozen + getGraph().nodes matches
 * Adding a 4th node = immediate L0 red-line violation.
 * Renaming any node = immediate L0 red-line violation.
 *
 * ─── §C3 Secret ban ───────────────────────────────────────────────────────────
 * ZERO process.env / getApiKey() / keychain.read("ark_api_key") access here.
 * All LLM calls go through deps.llm.generateObject(...) via the agent layer.
 *
 * ─── Channel-key vs node-name collision ───────────────────────────────────────
 * No rename needed. Channel keys (parse_payload, research_payload, direction_framework,
 * research_skipped, research_error, research_input, framework_config, resume_text,
 * jd_text, research_opt_in) are ALL different from node names (*_node suffix).
 * Unlike turn_graph's next_question_out rename, NO suffix workaround required here.
 */

import { StateGraph, START, END } from "@langchain/langgraph";
import type { StateGraphArgs } from "@langchain/langgraph";
import {
  ResearchAgentInputSchema,
  type ResearchAgentInput,
  type ResearchAgentOutput,
} from "@/core/schemas/research";
import {
  type FrameworkConfigInput,
  type FrameworkAgentOutput,
} from "@/core/schemas/frameworks";
import { type ParseResultPayload } from "@/core/schemas/parse";
import type { LLMProvider } from "@/core/llm/types";
import { runParseAgent } from "@/core/agents/parse";
import { runResearchAgent } from "@/core/agents/research";
import { runFrameworkAgent } from "@/core/agents/framework";

// §C3: No process.env / getApiKey / keychain access. LLM calls go through deps.llm only.

// ─── L0 #13: Node-name lock — layer ① and ② ─────────────────────────────────
// Modifying this set = immediate L0 red-line violation.

export type IntakeGraphNodeName =
  | "parse_node"
  | "research_node"
  | "predict_questions_node";

export const INTAKE_GRAPH_NODES = Object.freeze(
  new Set(["parse_node", "research_node", "predict_questions_node"] as const),
);

// ─── Constants ───────────────────────────────────────────────────────────────

/** 15s hard timeout for research_node per M3.3.2 spec §5.2. */
export const RESEARCH_TIMEOUT_MS = 15_000;

// ─── IntakeState ─────────────────────────────────────────────────────────────
// TS type alias (NOT Zod schema) — matches turn_graph approach.
// Mirrors Python intake_graph.py IntakeState.

type IntakeState = {
  // ===== Inputs (set before invoke) =====
  resume_text: string;
  jd_text: string;
  research_opt_in: boolean;                         // default false
  research_input: ResearchAgentInput | null;        // backward-compat hatch (tests pre-build)
  framework_config: FrameworkConfigInput | null;    // null ⇒ predict_questions_node no-op

  // ===== Outputs (null until node populates) =====
  parse_payload: ParseResultPayload | null;
  research_payload: ResearchAgentOutput | null;
  research_skipped: boolean;                        // default false
  research_error: string | null;                    // error class name as string, or null
  direction_framework: FrameworkAgentOutput | null;
};

// ─── Channel reducers (§4) ────────────────────────────────────────────────────
// `reducer:` API — supported in LangGraph.js 0.2.74 (same as turnGraph.ts).
// Pattern: last-write-wins with null-passthrough (r ?? l).

const channels: StateGraphArgs<IntakeState>["channels"] = {
  resume_text:         { reducer: (l, r) => r ?? l, default: () => "" },
  jd_text:             { reducer: (l, r) => r ?? l, default: () => "" },
  research_opt_in:     { reducer: (l, r) => r ?? l, default: () => false },
  research_input:      { reducer: (l, r) => r ?? l, default: () => null },
  framework_config:    { reducer: (l, r) => r ?? l, default: () => null },
  parse_payload:       { reducer: (l, r) => r ?? l, default: () => null },
  research_payload:    { reducer: (l, r) => r ?? l, default: () => null },
  research_skipped:    { reducer: (l, r) => r ?? l, default: () => false },
  research_error:      { reducer: (l, r) => r ?? l, default: () => null },
  direction_framework: { reducer: (l, r) => r ?? l, default: () => null },
};

// ─── IntakeGraphDeps ──────────────────────────────────────────────────────────

export interface IntakeGraphDeps {
  llm: LLMProvider;
  logger?: {
    info?(msg: string, meta?: Record<string, unknown>): void;
    warn?(msg: string, meta?: Record<string, unknown>): void;
  };
}

// ─── Nodes ────────────────────────────────────────────────────────────────────

// §5.1 — parse_node: delegates to runParseAgent — no timeout, no soft-skip.
// Parse failure propagates (intake fails). Logger is omitted intentionally:
// runParseAgent's ParseAgentDeps.logger.warn(msg, ...args) uses variadic args,
// while IntakeGraphDeps.logger uses meta?: Record<string, unknown>. Passing
// { llm: deps.llm } only avoids the interface mismatch.
async function parseNode(
  state: IntakeState,
  deps: IntakeGraphDeps,
): Promise<Partial<IntakeState>> {
  const parse_output = await runParseAgent(
    { resume_text: state.resume_text, jd_text: state.jd_text },
    { llm: deps.llm },
  );
  return { parse_payload: parse_output };
}

// §5.2 — research_node: Promise.race + AbortController + Symbol sentinel + 15s timeout.
// 3-way soft-skip: opt_out | insufficient_jd_signal | timeout.
// Logging key lock: research_node_skipped / research_node_timeout / research_node_llm_error.
async function researchNode(
  state: IntakeState,
  deps: IntakeGraphDeps,
): Promise<Partial<IntakeState>> {
  // Step 1: pick research_input — backward-compat slot wins, else derive from parse
  let research_input: ResearchAgentInput | null = state.research_input;
  if (research_input === null) {
    research_input = deriveResearchInputFromParse(state);
  }

  // Step 2: opt-out / no signal ⇒ soft skip
  if (research_input === null) {
    const reason = state.research_opt_in ? "insufficient_jd_signal" : "opt_out";
    deps.logger?.info?.("research_node_skipped", { reason });
    return {
      research_payload: null,
      research_skipped: true,
      research_error: null,
    };
  }

  // Step 3: race runResearchAgent vs 15s timeout
  const controller = new AbortController();
  const timeoutSentinel = Symbol("research_timeout");
  const timeoutPromise = new Promise<typeof timeoutSentinel>((resolve) => {
    setTimeout(() => {
      controller.abort();
      resolve(timeoutSentinel);
    }, RESEARCH_TIMEOUT_MS);
  });

  try {
    const result = await Promise.race([
      runResearchAgent(research_input, { llm: deps.llm }),
      timeoutPromise,
    ]);
    if (result === timeoutSentinel) {
      deps.logger?.info?.("research_node_timeout");
      return {
        research_payload: null,
        research_skipped: true,
        research_error: "timeout",
      };
    }
    return {
      research_payload: result,
      research_skipped: false,
      research_error: null,
    };
  } catch (err) {
    const errorName = (err as Error)?.constructor?.name ?? "Error";
    deps.logger?.info?.("research_node_llm_error", { reason: errorName });
    return {
      research_payload: null,
      research_skipped: true,
      research_error: errorName,
    };
  }
}

// §5.3 — deriveResearchInputFromParse: exported for unit tests.
// §A11 PII guard: uses ResearchAgentInputSchema.safeParse to filter;
// never passes raw company_name / role_title to logs (cache_key hash used inside runResearchAgent).
export function deriveResearchInputFromParse(
  state: IntakeState,
): ResearchAgentInput | null {
  if (!state.research_opt_in) return null;
  const payload = state.parse_payload;
  if (payload === null) return null;
  const company = (payload.jd_company_name ?? "").trim();
  const role = (payload.jd_role_title ?? "").trim();
  if (!company || !role) return null;
  let hints = (payload.jd_industry_hints ?? [])
    .map((h) => h.trim())
    .filter((h) => h.length > 0);
  if (hints.length === 0) hints = [role];
  const result = ResearchAgentInputSchema.safeParse({
    company_name: company.slice(0, 80),
    role_title: role.slice(0, 80),
    industry_hints: hints.slice(0, 5),
  });
  return result.success ? result.data : null;
}

// §5.4 — predict_questions_node: double-skip guard + Framework + soft-skip on error.
// Logging key: predict_questions_node_framework_error + {reason: errorClassName}.
async function predictQuestionsNode(
  state: IntakeState,
  deps: IntakeGraphDeps,
): Promise<Partial<IntakeState>> {
  if (state.framework_config === null || state.parse_payload === null) {
    return { direction_framework: null };
  }

  const research_payload_json: string | null =
    state.research_payload !== null
      ? JSON.stringify(state.research_payload)
      : null;

  const framework_input = {
    parse_payload_json: JSON.stringify(state.parse_payload),
    config: state.framework_config,
    research_payload_json,
  };

  try {
    const framework_output = await runFrameworkAgent(framework_input, { llm: deps.llm });
    return { direction_framework: framework_output };
  } catch (err) {
    const errorName = (err as Error)?.constructor?.name ?? "Error";
    deps.logger?.info?.("predict_questions_node_framework_error", { reason: errorName });
    return { direction_framework: null };
  }
}

// ─── buildIntakeGraph factory ─────────────────────────────────────────────────
// §6.2 — 4 edges, strictly sequential, NO fan-out.
// START→parse_node→research_node→predict_questions_node→END

export function buildIntakeGraph(deps: IntakeGraphDeps) {
  const graph = new StateGraph<IntakeState>({ channels })
    .addNode("parse_node",             (s) => parseNode(s, deps))
    .addNode("research_node",          (s) => researchNode(s, deps))
    .addNode("predict_questions_node", (s) => predictQuestionsNode(s, deps))
    .addEdge(START, "parse_node")
    .addEdge("parse_node", "research_node")
    .addEdge("research_node", "predict_questions_node")
    .addEdge("predict_questions_node", END);
  return graph.compile();
}
